packages/@uppy/companion/src/server/provider/webdav/auth/index.js

```diff
@@ -0,0 +1,79 @@
+const { XMLParser } = require('fast-xml-parser')
+
+const WebdavProvider = require('../WebdavProvider')
+const { getProtectedGot, validateURL } = require('../../../helpers/request')
+
+const cloudTypePathMappings = {
+  nextcloud: {
+    manual_revoke_url: '/settings/user/security',
+  },
+  owncloud: {
+    manual_revoke_url: '/settings/personal?sectionid=security',
+  },
+}
+
+class WebdavAuth extends WebdavProvider {
+  constructor (options) {
+    super(options)
+    this.authProvider = WebdavAuth.authProvider
+  }
+
+  // for "grant"
+  static getExtraConfig () {
+    return {}
+  }
+
+  static get authProvider () {
+    return 'webdavAuth'
+  }
+
+  getBaseUrl ({ query: { subdomain } }) {
+    const { protocol } = this.providerOptions
+
+    return `${protocol}://${subdomain}`
+  }
+
+  async getUsername ({ token, query }) {
+    const { allowLocalUrls } = this
+
+    const url = `${this.getBaseUrl({ query })}/ocs/v1.php/cloud/user`
+    if (!validateURL(url, allowLocalUrls)) {
+      throw new Error('invalid user url')
+    }
+
+    const response = await getProtectedGot({ url, blockLocalIPs: !allowLocalUrls }).get(url, {
+      headers: {
+        Authorization: `Bearer ${token}`,
+      },
+    }).text()
```

Comment:
How to apply measurements against further attacks here? Huge files, slow servers,...